junit-jupiter-api/src/main/java/org/junit/jupiter/api/parallel/UseResource.java

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright 2015-2018 the original author or authors.
+ *
+ * All rights reserved. This program and the accompanying materials are
+ * made available under the terms of the Eclipse Public License v2.0 which
+ * accompanies this distribution and is available at
+ *
+ * http://www.eclipse.org/legal/epl-v20.html
+ */
+
+package org.junit.jupiter.api.parallel;
+
+import static org.apiguardian.api.API.Status.EXPERIMENTAL;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Repeatable;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+import org.apiguardian.api.API;
+
+/**
+ * {@code @UseResource} is used to declare that the annotated test class or test
+ * method requires access to a resource identified by a key.
+ *
+ * <p>The resource key is specified using {@link #value()}. In addition,
+ * {@link #mode()} allows to specify whether the annotated test class or test
+ * method requires {@link ResourceAccessMode#READ_WRITE} or only
+ * {@link ResourceAccessMode#READ} access to the resource. In the former case,
+ * execution of the annotated element will occur while no other test class or
+ * test method that uses this resource is being executed. In the latter case,
+ * the annotated element may be executed concurrently with other test classes or
+ * methods that also require {@link ResourceAccessMode#READ} access but not at
+ * the same time as any other test that requires
+ * {@link ResourceAccessMode#READ_WRITE} access.
+ *
+ * <p>This annotation can be repeated to declare the use of multiple resources.
+ *
+ * @see Resources
+ * @see ResourceAccessMode
+ * @see UseResources
+ * @since 5.3
+ */
+@API(status = EXPERIMENTAL, since = "5.3")
+@Retention(RetentionPolicy.RUNTIME)
+@Target({ ElementType.TYPE, ElementType.METHOD })
+@Repeatable(UseResources.class)
+public @interface UseResource {
```

Comment:
Wouldn't `@ResourceLock` be a better suited name?

Reply:
Possibly... let's discuss it in the next team call, shall we?

Reply:
Isn't the fact that it is currently implemented via a lock an implementation detail? The idea of `@UseResource` was so that the tests could declare what resources they use and what guarantees they expect/observe. If we change the implementation to use something else than a Lock then `@ResourceLock` wouldn't make sense.

Reply:
I think this is one of those scenarios where an implementation detail leaks to higher levels of the overall abstraction simply because it's part of the high-level use case.

To me, `@UseResource` is too generic in that it does not immediately provide a mental link to the concurrency model in question. After all, we're talking about executing things in parallel. So there is going to have to be some sort of _locking_ in order to avoid contention -- right?

On the other hand, I understand your desire not to leak implementation details, but I'm still not fond of `@UseResource`. 😉

Reply:
In terms of consistency, IIRC, the only places we use "verbs" for annotation names are for `@ExtendWith` and `@RegisterExtension`.

Otherwise, we try to avoid using verbs for annotation names.

Reply:
> Possibly... let's discuss it in the next team call, shall we?

If I'm there... gladly.

Reply:
If people don't like `@ResourceLock`, I think that `@SharedResource` would at least be better than `@UseResource`.

Rationale: the developer is not instructing the framework to _use_ a resource for the test. Rather, the developer is instructing the framework that the test itself uses a shared resource, to which the framework should then synchronize access.

Reply:
I have to agree with @sbrannen that `@UseResource` is not a good name for this. `@ResourceLock` would make more sense to me.

Adding the Lock concept does help me to understand the underlying mental(concurrency) model. 

When seeing for the first time `@UseResource` for some reason I imagined that this annotation would be added to some `resource` ie some critical piece of code eg a method, that would be shared and the framework would orchestrate tests around it. _[In retrospect, feeling a bit stupid now for thinking that.]_

After reviewing the code I think the important point of this mechanism is the exclusivity of execution of tests, ie these tests characterized as _readers_ can execute at the same time while tests characterized as _writers_ can only run by themselves.

The concept of a lock suggests the above quite well for me.